shared/health/src/test/java/io/pravega/shared/health/HealthServiceUpdaterTests.java

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.shared.health;
+
+import lombok.Cleanup;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import io.pravega.shared.health.TestHealthContributors.HealthyContributor;
+import io.pravega.shared.health.TestHealthContributors.FailingContributor;
+
+import io.pravega.test.common.AssertExtensions;
+
+@Slf4j
+public class HealthServiceUpdaterTests {
+
+    HealthService service;
+
+    HealthServiceUpdater healthServiceUpdater;
+
+    HealthServiceFactory factory;
+    @Before
+    public void before() {
+        factory = new HealthServiceFactory();
+        service = factory.createHealthService("health-service");
+        service.getHealthServiceUpdater().startAsync();
+        healthServiceUpdater = service.getHealthServiceUpdater();
+        healthServiceUpdater.awaitRunning();
+    }
+
+    @After
+    public void after() {
+        factory.close();
+        service.close();
+        healthServiceUpdater.stopAsync();
+        healthServiceUpdater.awaitTerminated();
```

Comment:
You need to invoke `.close()` on healthServiceUpdater. If you look in your implementation, this is the only place where you shut down the executor it creates; otherwise it will leave threads hanging around for the remainder of your process.

Please verify your other unit tests as well - it is not exactly easy to trace this on Github code review.

Reply:
`close` is invoked for the healthServiceUpdater from within the `close` call of the service.